### PR TITLE
[SPARK-41203] [CONNECT] Support Dataframe.tansform in Python client.

### PR DESCRIPTION
**Pull request by grundprinzip**
### What changes were proposed in this pull request?
Support client-side transformations of the DataFrame.

### Why are the changes needed?
Compatibility

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT

### CODE REVIEW

**Comment by amaliujia** [comment]
Please always add `[CONNECT]` in PR title so it is easy to filter out all connect commits through `git log`.

**Comment by grundprinzip** [comment]
> Please always add `[CONNECT]` in PR title so it is easy to filter out all connect commits through `git log`.

Thanks. I missed it. Fixed

File: python/pyspark/sql/connect/dataframe.py [hunk 1]
```diff
@@ -756,6 +757,62 @@ def schema(self) -> StructType:
         else:
             return self._schema
 
+    def transform(self, func: Callable[..., "DataFrame"], *args: Any, **kwargs: Any) -> "DataFrame":
+        """Returns a new :class:`DataFrame`. Concise syntax for chaining custom transformations.
+
+        .. versionadded:: 3.0.0
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/dataframe.py:763]:
```suggestion

        .. versionadded:: 3.3.0
```

File: python/pyspark/sql/connect/dataframe.py [hunk 2]
```diff
@@ -756,6 +757,62 @@ def schema(self) -> StructType:
...
+
+        Parameters
+        ----------
+        func : function
+            a function that takes and returns a :class:`DataFrame`.
+        *args
+            Positional arguments to pass to func.
+
+            .. versionadded:: 3.3.0
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/dataframe.py:772]:
```suggestion
```

File: python/pyspark/sql/connect/dataframe.py [hunk 3]
```diff
@@ -756,6 +757,62 @@ def schema(self) -> StructType:
...
+        **kwargs
+            Keyword arguments to pass to func.
+
+            .. versionadded:: 3.3.0
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/dataframe.py:776]:
```suggestion
```

**Review by HyukjinKwon — APPROVED** [review]
LGTM, just some nits

File: python/pyspark/sql/connect/dataframe.py (hunk 1 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/dataframe.py:763]: fixed

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**